content/content/oop.md

```diff
@@ -4,9 +4,108 @@ title: Object Oriented Programming
 
 # Object Oriented Programming
 
-Object oriented programming is supported through inheritance and methods, which are virtual procedures. Methods work in the same way as [procs](/procs/), including UFCS, but the runtime type of the object that a method is called with is used to determine which version to call.
+## Member functions or Method calls
 
-Inheritance is created with the `of` keyword in the type declaration of an object. A method is overridden by creating a new method with parameter types of the subtype. 
+Member functions, typically known as methods in OOP terminology, can be implemented directly using `proc`s, thanks to the Uniform Call Syntax.
+
+``` nimrod
+type Animal = object
+  name: string
+  age: int
+  
+proc speak(self: Animal, msg: string) = 
+  echo self.name & " says:" & msg 
+  
+var sparky = Animal(name: "Sparky", age: 10)
+
+sparky.speak("Hi")
+# this is the same as the below
+speak(sparky, "Hi")
+```
+
+As you might have guessed, the method call syntax is not restricted to object types.
+
+``` nimrod
+proc double(num: int):int =
+  return num*2
+  
+double(10)
+# same as
+10.double()
+# you can even drop the empty paranthesis
+10.double
+```
+
+Now, when using UFCS to think about `proc`s as member functions, there is an important point to keep in mind about mutability. Consider this example.
+
+``` nimrod
+type Animal = object
+  name: string
+  age: int
+  
+proc incAge(self: Animal) =
+  self.age += 1
+  
+var sparky = Animal(name: "Sparky", age: 3)
+```
+
+This fails compiling with the following error:
+``` console
+Error: type mismatch: got <int, int literal(1)>
+but expected one of:
+proc `+=`[T: SomeInteger](x: var T; y: T)
+  first type mismatch at position: 1
+  required type for x: var T: SomeInteger
+  but expression 'self.age' is immutable, not 'var'
+proc `+=`[T: float | float32 | float64](x: var T; y: T)
+  first type mismatch at position: 1
+  required type for x: var T: float or float32 or float64
+  but expression 'self.age' is of type: int
+
+expression: self.age += 1
+```
```

Comment:
I would remove these--I don't think the more complex error shows anything that the simpler one doesn't.

Reply:
An even more inscrutable (to a beginner like me, at least) error message (https://github.com/nim-lang/Nim/issues/20779) was what got me to writing this section.

May be, I can remove this example and instead just have a note on function type mismatch errors in these scenarios.

Reply:
Ok, I see where you're coming from. I think it's reasonable to keep--but that the simple example should come first to familiarize the reader with the situation. I think it'd also be a good idea to remove the second half of the error message, about the "float"s.